editor/plugins/canvas_item_editor_plugin.cpp

```diff
@@ -5743,47 +5757,69 @@ void CanvasItemEditorViewport::_perform_drop_data() {
-
-			for (int i = 0; i < files.size(); i++) {
-				String extension = files[i].get_extension().to_lower();
-
-				// Check if dragged files with texture or scene extension can be created at least once.
-				if (texture_extensions.find(extension) || scene_extensions.find(extension)) {
-					Ref<Resource> res = ResourceLoader::load(files[i]);
-					if (res.is_null()) {
-						continue;
-					}
-					Ref<PackedScene> scn = res;
-					if (scn.is_valid()) {
-						Node *instantiated_scene = scn->instantiate(PackedScene::GEN_EDIT_STATE_INSTANCE);
-						if (!instantiated_scene) {
-							continue;
-						}
-						memdelete(instantiated_scene);
-					}
+	if (d.has("type") && String(d["type"]) == "files") {
+		Vector<String> files = d["files"];
+		bool can_instantiate = false;
+		String drop_type;
+
+		List<String> scene_extensions;
+		ResourceLoader::get_recognized_extensions_for_type("PackedScene", &scene_extensions);
+		List<String> texture_extensions;
+		ResourceLoader::get_recognized_extensions_for_type("Texture2D", &texture_extensions);
+		List<String> audio_extensions;
+		ResourceLoader::get_recognized_extensions_for_type("AudioStream", &texture_extensions);
+
+		for (int i = 0; i < files.size(); i++) {
+			String extension = files[i].get_extension().to_lower();
+
+			// Optimization to not load a resource every time.
+			if (!(texture_extensions.find(extension) || scene_extensions.find(extension) || audio_extensions.find(extension))) {
+				continue;
+			}
+			// Check if dragged files with texture or scene extension can be created at least once.
+			Ref<Resource> res = ResourceLoader::load(files[i]);
+			if (res.is_null()) {
+				continue;
+			}
+			String type = res->get_class();
+			if (type == "PackedScene") {
+				Ref<PackedScene> sdata = Object::cast_to<PackedScene>(*res);
+				Node *instantiated_scene = sdata->instantiate(PackedScene::GEN_EDIT_STATE_INSTANCE);
+				if (instantiated_scene) {
+					memdelete(instantiated_scene);
+					drop_type = "scene";
 					can_instantiate = true;
 					break;
 				}
-			}
-			if (can_instantiate) {
-				if (!preview_node->get_parent()) { // create preview only once
-					_create_preview(files);
+			} else if (ClassDB::is_parent_class(type, "Texture2D")) {
+				Ref<Texture2D> texture = Object::cast_to<Texture2D>(*res);
+				if (texture.is_valid()) {
+					drop_type = "texture";
+					can_instantiate = true;
+					break;
+				}
+			} else if (ClassDB::is_parent_class(type, "AudioStream")) {
+				Ref<AudioStream> audio_stream = Object::cast_to<AudioStream>(*res);
+				if (audio_stream.is_valid()) {
+					drop_type = "audio_stream";
+					can_instantiate = true;
+					break;
 				}
-				Transform2D trans = canvas_item_editor->get_canvas_transform();
-				preview_node->set_position((p_point - trans.get_origin()) / trans.get_scale().x);
+			}
+		}
+		if (can_instantiate) {
+			if (!preview_node->get_parent()) { // create preview only once
```

Comment:
```suggestion
			if (!preview_node->get_parent()) { // Create preview only once.
```
Per the [comment style](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide), applies to copied or rearranged code as well